locust/runners.py

```diff
@@ -189,25 +191,38 @@ def stop_users(self, user_count):
                     to_stop.append(user)
                     bucket.remove(l)
                     break
-        self.stop_user_instances(to_stop)
-        self.environment.events.hatch_complete.fire(user_count=self.user_count)
-    
-    
-    def stop_user_instances(self, users):
-        if self.environment.stop_timeout:
-            stopping = Group()
-            for user in users:
-                if not user.stop(self.user_greenlets, force=False):
+
+        if not to_stop:
+            return
+
+        if stop_rate == None or stop_rate >= user_count:
+            sleep_time = 0
+            logger.info("Stopping %i users immediately" % (user_count))
+        else:
+            sleep_time = 1.0 / stop_rate
+            logger.info("Stopping %i users at rate of %g users/s" % (user_count, stop_rate))
+
+        while True:
+            user_to_stop = to_stop.pop(random.randint(0, len(to_stop)-1))
+            logger.debug('Stopping %s' % user_to_stop._greenlet.name)
+            if self.environment.stop_timeout:
+                stop_group = Group()
+                if not user_to_stop.stop(self.user_greenlets, force=False):
                     # User.stop() returns False if the greenlet was not stopped, so we'll need
                     # to add it's greenlet to our stopping Group so we can wait for it to finish it's task
-                    stopping.add(user._greenlet)
-            if not stopping.join(timeout=self.environment.stop_timeout):
-                logger.info("Not all users finished their tasks & terminated in %s seconds. Stopping them..." % self.environment.stop_timeout)
-            stopping.kill(block=True)
-        else:
-            for user in users:
-                user.stop(self.user_greenlets, force=True)
-        
+                    stop_group.add(user_to_stop._greenlet)
+                if not stop_group.join(timeout=self.environment.stop_timeout):
```

Comment:
This timeout will prolong the iteration time for this while loop. 

So if you had for example stop_rate 1 and stop_timeout 2, you would end up (if no users stopping within the timeout) taking 3 seconds to complete an iteration (giving you a stop rate of 1/3 users per second). As a user, that is not what I would expect. Ideally these things should be done async (I guess you'll have to spawn a "killer greenlet" or something)

Reply:
OK I see what you mean.

What is we just do `stop_group = Group()` outside the loop and then `stop_group.kill(block=True)` after loop is done?

Reply:
You mean *inside* the loop, right? Sounds good! Can you add a test case for it?

Reply:
Sounds good! Can you add a test for it? I think it is one of those things that could trip us up :) (now or in the future)

Reply:
I can but what exactly is the test? Ramping down with `stop_timeout` set? How does the test fail? If ramping down doesn't happen? Or it happens too slowly?

If I understand correctly, if `stop_timeout` is set, we can't really guarantee any ramp down rate accurately, but could do best effort.